File: max_water_container.cpp
//Time comlpexity - Fine below
//Space complexity - Find below
//Ran successfully on leetcode
//3 point approach is the following -
//1.We assign 2 pointers i.e. low and high
//2.We calculate the area of the boundary formed by the two pointers.
//3.We compare the lengths of the two pointers and move away from the low one.

class Solution {
public:
    int maxArea(vector<int>& height) {
        if(height.size()==0)
            return 0;
        /*Using brute force O(n^2) time and O(1) space
        int m=0;
        for(int i=0;i<height.size()-1;i++)
            for(int j=i+1;j<height.size();j++)
            {
                m=max(m,(j-i)*min(height[i],height[j]));
            }
        return m;*/
        
        //Using two pointers with O(n) time and O(1) space
        int l=0;
        int h=height.size()-1;
        int m=0; //max
        while(l<h)
        {
            int current_area= min(height[l],height[h])*(h-l);
            if(current_area>m)
                m=current_area;
            if(height[l]<height[h])
                l++;
            else
                h--;
        }
        return m;
    }
};

File: MaxWaterContainer.java

/**
 * Take two pointers left and right. Take the min of them
 * and multiply with diff of their indices. Update max value
 * if the current product is maximum. Then to move pointers,
 * we check if the left pointer height is less than right pointer
 * height. If so do l++, else h--;
 *
 */
class Solution {
	public int maxArea(int[] height) {
		int l = 0, h = height.length - 1;
		int max = 0;
		while (l < h) {
			int min = Math.min(height[l], height[h]);
			max = Math.max(max, (h - l) * min);
			if (height[l] < height[h])
				l++;
			else
				h--;
		}
		return max;
	}
}